network/internal/p2pfixtures/fixtures.go

```diff
@@ -454,7 +480,257 @@ func LetNodesDiscoverEachOther(t *testing.T, ctx context.Context, nodes []p2p.Li
 			require.NoError(t, node.AddPeer(ctx, otherPInfo))
 		}
 	}
+}
+
+// AddNodesToEachOthersPeerStore adds the dialing address of all nodes to the peer store of all other nodes.
+// However, it does not connect them to each other.
+func AddNodesToEachOthersPeerStore(t *testing.T, nodes []p2p.LibP2PNode, ids flow.IdentityList) {
+	for _, node := range nodes {
+		for i, other := range nodes {
+			if node == other {
+				continue
+			}
+			otherPInfo, err := utils.PeerAddressInfo(*ids[i])
+			require.NoError(t, err)
+			node.Host().Peerstore().AddAddrs(otherPInfo.ID, otherPInfo.Addrs, peerstore.AddressTTL)
+		}
+	}
+}
+
+// EnsureConnected ensures that the given nodes are connected to each other.
+// It fails the test if any of the nodes is not connected to any other node.
+func EnsureConnected(t *testing.T, ctx context.Context, nodes []p2p.LibP2PNode) {
+	for _, node := range nodes {
+		for _, other := range nodes {
+			if node == other {
+				continue
+			}
+			require.NoError(t, node.Host().Connect(ctx, other.Host().Peerstore().PeerInfo(other.Host().ID())))
+		}
+	}
+}
 
-	// wait for all nodes to discover each other
-	time.Sleep(time.Second)
+// EnsureNotConnected ensures that no connection exists from "from" nodes to "to" nodes.
+func EnsureNotConnected(t *testing.T, ctx context.Context, from []p2p.LibP2PNode, to []p2p.LibP2PNode) {
+	for _, node := range from {
+		for _, other := range to {
+			if node == other {
+				require.Fail(t, "overlapping nodes in from and to lists")
+			}
+			require.Error(t, node.Host().Connect(ctx, other.Host().Peerstore().PeerInfo(other.Host().ID())))
+		}
+	}
+}
+
+// EnsureNotConnectedBetweenGroups ensures no connection exists between the given groups of nodes.
+func EnsureNotConnectedBetweenGroups(t *testing.T, ctx context.Context, groupA []p2p.LibP2PNode, groupB []p2p.LibP2PNode) {
+	// ensure no connection from group A to group B
+	EnsureNotConnected(t, ctx, groupA, groupB)
+	// ensure no connection from group B to group A
+	EnsureNotConnected(t, ctx, groupB, groupA)
+}
+
+// EnsurePubsubMessageExchange ensures that the given nodes exchange the given message on the given channel through pubsub.
+func EnsurePubsubMessageExchange(t *testing.T, ctx context.Context, nodes []p2p.LibP2PNode, messageFactory func() (interface{}, channels.Topic)) {
+	_, topic := messageFactory()
+
+	subs := make([]*pubsub.Subscription, len(nodes))
+	slashingViolationsConsumer := unittest.NetworkSlashingViolationsConsumer(unittest.Logger(), metrics.NewNoopCollector())
+	var err error
+	for i, node := range nodes {
+		subs[i], err = node.Subscribe(
+			topic,
+			validator.TopicValidator(
+				unittest.Logger(),
+				unittest.NetworkCodec(),
+				slashingViolationsConsumer,
+				unittest.AllowAllPeerFilter()))
+		require.NoError(t, err)
+	}
+
+	// let subscriptions propagate
+	time.Sleep(1 * time.Second)
```

Comment:
Have you seen any flakiness using this sleep method?

Reply:
None so far.